File: aqi-site/aqi_site/alerts.py
"""Alerting: evaluate the mask level per sensor and push to ntfy / a webhook.

The warehouse stays read-only; alert state (last level, last-fired time) and runtime
overrides live in the sidecar's own SQLite database. Firing uses hysteresis (only on
escalation or after a min interval) and honours quiet hours so it will not spam.
"""
from __future__ import annotations

import json
import logging
import sqlite3
import threading
import time
from dataclasses import replace
from pathlib import Path

try:  # requests ships with the pm25 core deps; guard so imports never fail
    import requests
except Exception:  # pragma: no cover
    requests = None  # type: ignore[assignment]

from . import analytics, db, mask
from .config import AlertsConfig, Config

log = logging.getLogger("aqi_site.alerts")

_STATE_SCHEMA = (
    "CREATE TABLE IF NOT EXISTS alert_state ("
    "sensor_id TEXT PRIMARY KEY, last_level TEXT NOT NULL DEFAULT 'none', "
    "last_fired_ts INTEGER NOT NULL DEFAULT 0)",
    "CREATE TABLE IF NOT EXISTS alert_overrides ("
    "id INTEGER PRIMARY KEY CHECK (id=1), data TEXT NOT NULL)",
)

_OVERRIDE_KEYS = {
    "enabled",
    "notify_from",
    "min_interval_s",
    "quiet_start_hour",
    "quiet_end_hour",
}

_PRIORITY = {"indoors": "urgent", "strong": "high", "recommended": "default"}
_TAGS = {"indoors": "no_entry", "strong": "rotating_light", "recommended": "mask"}


def open_state(path: str) -> sqlite3.Connection:
    Path(path).parent.mkdir(parents=True, exist_ok=True)
    conn = sqlite3.connect(path, timeout=10)
    conn.row_factory = sqlite3.Row
    conn.execute("PRAGMA journal_mode=WAL")
    for statement in _STATE_SCHEMA:
        conn.execute(statement)
    conn.commit()
    return conn


def get_state(conn: sqlite3.Connection, sensor_id: str) -> tuple[str, int]:
    row = conn.execute(
        "SELECT last_level, last_fired_ts FROM alert_state WHERE sensor_id=?",
        (sensor_id,),
    ).fetchone()
    if row is None:
        return "none", 0
    return row["last_level"], row["last_fired_ts"]


def set_state(conn: sqlite3.Connection, sensor_id: str, level: str, fired_ts: int) -> None:
    conn.execute(
        "INSERT INTO alert_state (sensor_id, last_level, last_fired_ts) VALUES (?, ?, ?) "
        "ON CONFLICT(sensor_id) DO UPDATE SET last_level=excluded.last_level, "
        "last_fired_ts=excluded.last_fired_ts",
        (sensor_id, level, fired_ts),
    )
    conn.commit()


def load_overrides(conn: sqlite3.Connection) -> dict:
    row = conn.execute("SELECT data FROM alert_overrides WHERE id=1").fetchone()
    if row is None:
        return {}
    try:
        data = json.loads(row["data"])
    except (json.JSONDecodeError, TypeError):
        return {}
    return {k: v for k, v in data.items() if k in _OVERRIDE_KEYS} if isinstance(data, dict) else {}


def save_overrides(conn: sqlite3.Connection, data: dict) -> dict:
    clean = {k: data[k] for k in _OVERRIDE_KEYS if k in data}
    conn.execute(
        "INSERT INTO alert_overrides (id, data) VALUES (1, ?) "
        "ON CONFLICT(id) DO UPDATE SET data=excluded.data",
        (json.dumps(clean),),
    )
    conn.commit()
    return clean


def effective_alerts(base: AlertsConfig, overrides: dict) -> AlertsConfig:
    if not overrides:
        return base
    return replace(
        base,
        enabled=bool(overrides.get("enabled", base.enabled)),
        notify_from=str(overrides.get("notify_from", base.notify_from)).lower(),
        min_interval_s=int(overrides.get("min_interval_s", base.min_interval_s)),
        quiet_start_hour=int(overrides.get("quiet_start_hour", base.quiet_start_hour)),
        quiet_end_hour=int(overrides.get("quiet_end_hour", base.quiet_end_hour)),
    )


def in_quiet_hours(now: int, cfg: AlertsConfig, offset: int = 0) -> bool:
    if cfg.quiet_start_hour < 0 or cfg.quiet_end_hour < 0:
        return False
    hour = ((now + offset) // 3600) % 24
    start, end = cfg.quiet_start_hour, cfg.quiet_end_hour
    if start == end:
        return False
    if start < end:
        return start <= hour < end
    return hour >= start or hour < end


def should_fire(
    prev_level: str, new_level: str, cfg: AlertsConfig, last_fired_ts: int, now: int
) -> tuple[bool, str]:
    if mask.severity(new_level) < mask.severity(cfg.notify_from):
        return False, "below-threshold"
    if mask.severity(new_level) > mask.severity(prev_level):
        return True, "escalated"
    if now - last_fired_ts >= cfg.min_interval_s:
        return True, "reminder"
    return False, "debounced"


def send_ntfy(cfg: AlertsConfig, title: str, message: str, level: str) -> bool:
    if requests is None or not cfg.ntfy_url or not cfg.ntfy_topic:
        return False
    headers = {
        "Title": title,
        "Priority": _PRIORITY.get(level, "default"),
        "Tags": _TAGS.get(level, "mask"),
    }
    if cfg.ntfy_token:
        headers["Authorization"] = f"Bearer {cfg.ntfy_token}"
    try:
        resp = requests.post(
            f"{cfg.ntfy_url}/{cfg.ntfy_topic}",
            data=message.encode("utf-8"),
            headers=headers,
            timeout=10,
        )
        return resp.ok
    except Exception as exc:  # network failures must not crash the poller
        log.warning("ntfy push failed: %s", exc)
        return False


def send_webhook(cfg: AlertsConfig, payload: dict) -> bool:
    if requests is None or not cfg.webhook_url:
        return False
    try:
        resp = requests.post(cfg.webhook_url, json=payload, timeout=10)
        return resp.ok
    except Exception as exc:
        log.warning("webhook push failed: %s", exc)
        return False


def notify(cfg: AlertsConfig, sensor_id: str, snapshot: dict) -> bool:
    reco = snapshot["mask"]
    aqi_value = snapshot["nowcast_aqi"]
    title = f"{reco['label']} — AQI {aqi_value}"
    message = f"[{sensor_id}] {reco['action']} (dominant {snapshot['dominant']})."
    payload = {
        "sensor_id": sensor_id,
        "aqi": aqi_value,
        "level": reco["level"],
        "label": reco["label"],
        "action": reco["action"],
        "ts": snapshot["ts"],
    }
    sent_ntfy = send_ntfy(cfg, title, message, reco["level"])
    sent_webhook = send_webhook(cfg, payload)
    return sent_ntfy or sent_webhook


def poll_once(config: Config, state_conn: sqlite3.Connection) -> list[dict]:
    """Evaluate every sensor once and fire alerts as needed. Returns per-sensor events."""
    cfg = effective_alerts(config.alerts, load_overrides(state_conn))
    if not cfg.enabled:
        return []
    offset = int(round(config.display.tz_offset_hours * 3600))
    now = int(time.time())
    events: list[dict] = []
    try:
        conn = db.connect(config.db_path, config.db_immutable)
    except sqlite3.OperationalError as exc:
        log.warning("alert poll skipped, warehouse unavailable: %s", exc)
        return []
    try:
        for sensor_id in db.distinct_sensors(conn):
            snapshot = analytics.current(conn, sensor_id, config.mask, offset)
            if snapshot is None:
                continue
            level = snapshot["mask"]["level"]
            prev_level, last_fired = get_state(state_conn, sensor_id)
            fire, reason = should_fire(prev_level, level, cfg, last_fired, now)
            quiet = in_quiet_hours(now, cfg, offset)
            sent = False
            if fire and not quiet:
                sent = notify(cfg, sensor_id, snapshot)
            set_state(state_conn, sensor_id, level, now if sent else last_fired)
            if fire or level != prev_level:
                events.append(
                    {
                        "sensor_id": sensor_id,
                        "level": level,
                        "prev_level": prev_level,
                        "reason": "quiet" if (fire and quiet) else reason,
                        "sent": sent,
                    }
                )
    finally:
        conn.close()
    return events


def run_poller(config: Config, stop_event: threading.Event) -> None:
    """Background loop: poll on an interval until ``stop_event`` is set."""
    state_conn = open_state(config.state_db_path)
    try:
        while not stop_event.is_set():
            try:
                poll_once(config, state_conn)
            except Exception as exc:  # keep the thread alive across unexpected errors
                log.warning("alert poll error: %s", exc)
            stop_event.wait(config.alerts.poll_interval_s)
    finally:
        state_conn.close()

File: aqi-worker/src/lib/alerts.ts
// Alerting: evaluate the mask level per sensor and push to ntfy / a webhook. Runs from
// the Worker's Cron Trigger. Alert state (last level, last-fired) and runtime overrides
// live in D1. Firing uses hysteresis + quiet hours + a min interval so it won't spam.
import * as analytics from "./analytics.ts";
import type { AlertsConfig, Config } from "./config.ts";
import * as db from "./db.ts";
import type { Env } from "../env.ts";
import * as mask from "./mask.ts";

const OVERRIDE_KEYS = [
  "enabled",
  "notify_from",
  "min_interval_s",
  "quiet_start_hour",
  "quiet_end_hour",
] as const;

const PRIORITY: Record<string, string> = {
  indoors: "urgent",
  strong: "high",
  recommended: "default",
};
const TAGS: Record<string, string> = {
  indoors: "no_entry",
  strong: "rotating_light",
  recommended: "mask",
};

export async function getState(d: D1Database, sensorId: string): Promise<[string, number]> {
  const row = await d
    .prepare("SELECT last_level, last_fired_ts FROM alert_state WHERE sensor_id=?")
    .bind(sensorId)
    .first<{ last_level: string; last_fired_ts: number }>();
  if (row === null) return ["none", 0];
  return [row.last_level, row.last_fired_ts];
}

export async function setState(
  d: D1Database,
  sensorId: string,
  level: string,
  firedTs: number,
): Promise<void> {
  await d
    .prepare(
      "INSERT INTO alert_state (sensor_id, last_level, last_fired_ts) VALUES (?, ?, ?) " +
        "ON CONFLICT(sensor_id) DO UPDATE SET last_level=excluded.last_level, " +
        "last_fired_ts=excluded.last_fired_ts",
    )
    .bind(sensorId, level, firedTs)
    .run();
}

export async function loadOverrides(d: D1Database): Promise<Record<string, unknown>> {
  const row = await d
    .prepare("SELECT data FROM alert_overrides WHERE id=1")
    .first<{ data: string }>();
  if (row === null) return {};
  let data: unknown;
  try {
    data = JSON.parse(row.data);
  } catch {
    return {};
  }
  if (typeof data !== "object" || data === null) return {};
  const out: Record<string, unknown> = {};
  for (const k of OVERRIDE_KEYS) {
    if (k in (data as Record<string, unknown>)) out[k] = (data as Record<string, unknown>)[k];
  }
  return out;
}

export async function saveOverrides(
  d: D1Database,
  data: Record<string, unknown>,
): Promise<Record<string, unknown>> {
  const clean: Record<string, unknown> = {};
  for (const k of OVERRIDE_KEYS) {
    if (k in data) clean[k] = data[k];
  }
  await d
    .prepare(
      "INSERT INTO alert_overrides (id, data) VALUES (1, ?) " +
        "ON CONFLICT(id) DO UPDATE SET data=excluded.data",
    )
    .bind(JSON.stringify(clean))
    .run();
  return clean;
}

export function effectiveAlerts(
  base: AlertsConfig,
  overrides: Record<string, unknown>,
): AlertsConfig {
  if (Object.keys(overrides).length === 0) return base;
  const asInt = (v: unknown, fallback: number) =>
    typeof v === "number" ? Math.trunc(v) : fallback;
  return {
    ...base,
    enabled: "enabled" in overrides ? Boolean(overrides.enabled) : base.enabled,
    notify_from:
      "notify_from" in overrides ? String(overrides.notify_from).toLowerCase() : base.notify_from,
    min_interval_s: asInt(overrides.min_interval_s, base.min_interval_s),
    quiet_start_hour: asInt(overrides.quiet_start_hour, base.quiet_start_hour),
    quiet_end_hour: asInt(overrides.quiet_end_hour, base.quiet_end_hour),
  };
}

export function inQuietHours(nowTs: number, cfg: AlertsConfig, offset = 0): boolean {
  if (cfg.quiet_start_hour < 0 || cfg.quiet_end_hour < 0) return false;
  const hour = Math.floor((nowTs + offset) / 3600) % 24;
  const start = cfg.quiet_start_hour;
  const end = cfg.quiet_end_hour;
  if (start === end) return false;
  if (start < end) return start <= hour && hour < end;
  return hour >= start || hour < end;
}

export function shouldFire(
  prevLevel: string,
  newLevel: string,
  cfg: AlertsConfig,
  lastFiredTs: number,
  nowTs: number,
): [boolean, string] {
  if (mask.severity(newLevel) < mask.severity(cfg.notify_from)) return [false, "below-threshold"];
  if (mask.severity(newLevel) > mask.severity(prevLevel)) return [true, "escalated"];
  if (nowTs - lastFiredTs >= cfg.min_interval_s) return [true, "reminder"];
  return [false, "debounced"];
}

export async function sendNtfy(
  cfg: AlertsConfig,
  title: string,
  message: string,
  level: string,
): Promise<boolean> {
  if (!cfg.ntfy_url || !cfg.ntfy_topic) return false;
  const headers: Record<string, string> = {
    Title: title,
    Priority: PRIORITY[level] ?? "default",
    Tags: TAGS[level] ?? "mask",
  };
  if (cfg.ntfy_token) headers.Authorization = `Bearer ${cfg.ntfy_token}`;
  try {
    const resp = await fetch(`${cfg.ntfy_url}/${cfg.ntfy_topic}`, {
      method: "POST",
      body: message,
      headers,
    });
    return resp.ok;
  } catch {
    return false;
  }
}

export async function sendWebhook(
  cfg: AlertsConfig,
  payload: Record<string, unknown>,
): Promise<boolean> {
  if (!cfg.webhook_url) return false;
  try {
    const resp = await fetch(cfg.webhook_url, {
      method: "POST",
      body: JSON.stringify(payload),
      headers: { "content-type": "application/json" },
    });
    return resp.ok;
  } catch {
    return false;
  }
}

export async function notify(
  cfg: AlertsConfig,
  sensorId: string,
  snapshot: analytics.Current,
): Promise<boolean> {
  const reco = snapshot.mask;
  const aqiValue = snapshot.nowcast_aqi;
  const title = `${reco.label} — AQI ${aqiValue}`;
  const message = `[${sensorId}] ${reco.action} (dominant ${snapshot.dominant}).`;
  const payload = {
    sensor_id: sensorId,
    aqi: aqiValue,
    level: reco.level,
    label: reco.label,
    action: reco.action,
    ts: snapshot.ts,
  };
  const sentNtfy = await sendNtfy(cfg, title, message, reco.level);
  const sentWebhook = await sendWebhook(cfg, payload);
  return sentNtfy || sentWebhook;
}

export interface AlertEvent {
  sensor_id: string;
  level: string;
  prev_level: string;
  reason: string;
  sent: boolean;
}

/** Evaluate every sensor once and fire alerts as needed. Returns per-sensor events. */
export async function pollOnce(env: Env, config: Config): Promise<AlertEvent[]> {
  const cfg = effectiveAlerts(config.alerts, await loadOverrides(env.DB));
  if (!cfg.enabled) return [];
  const offset = config.offset_s;
  const nowTs = Math.floor(Date.now() / 1000);
  const events: AlertEvent[] = [];
  for (const sensorId of await db.distinctSensors(env.DB)) {
    const snapshot = await analytics.current(env.DB, sensorId, config.mask, offset);
    if (snapshot === null) continue;
    const level = snapshot.mask.level;
    const [prevLevel, lastFired] = await getState(env.DB, sensorId);
    const [fire, reason] = shouldFire(prevLevel, level, cfg, lastFired, nowTs);
    const quiet = inQuietHours(nowTs, cfg, offset);
    let sent = false;
    if (fire && !quiet) sent = await notify(cfg, sensorId, snapshot);
    await setState(env.DB, sensorId, level, sent ? nowTs : lastFired);
    if (fire || level !== prevLevel) {
      events.push({
        sensor_id: sensorId,
        level,
        prev_level: prevLevel,
        reason: fire && quiet ? "quiet" : reason,
        sent,
      });
    }
  }
  return events;
}
